DatadogInternal/Sources/NetworkInstrumentation/URLSession/URLSessionTaskInterception.swift

```diff
@@ -97,6 +97,30 @@ public struct ResourceCompletion {
     }
 }
 
+/// An immutable version of `URLRequest`.
+///
+/// Introduced in response to concerns raised in https://github.com/DataDog/dd-sdk-ios/issues/1638 
+/// it makes a copy of request attributes, safeguarding against potential thread safety issues arising from concurrent 
+/// mutations (see more context in https://github.com/DataDog/dd-sdk-ios/pull/1767 ).
+public struct ImmutableRequest {
```

Comment:
does this have to be public?

Reply:
It has `public` requirement [in `URLSessionTaskInterception`](https://github.com/DataDog/dd-sdk-ios/blob/5334eab9cb5f9de2f47f62d9a951b3b05230938e/DatadogInternal/Sources/NetworkInstrumentation/URLSession/URLSessionTaskInterception.swift#L12-L14) which is later leveraged in:
- `DatadogRUM`: [example](https://github.com/DataDog/dd-sdk-ios/blob/5334eab9cb5f9de2f47f62d9a951b3b05230938e/DatadogRUM/Sources/Instrumentation/Resources/URLSessionRUMResourcesHandler.swift#L74)
- `DatadogTrace`: [example](https://github.com/DataDog/dd-sdk-ios/blob/5334eab9cb5f9de2f47f62d9a951b3b05230938e/DatadogTrace/Sources/Integrations/TracingURLSessionHandler.swift#L132)

Unless we want to invest into bigger refactoring, I think there is no better way around. Keep in mind this is public in `DatadogInternal`, meaning not user-facing.